### PR TITLE
Version Management

### PR DESCRIPTION
**Pull request by mnm678**
Please fill in the fields below to submit a pull request.  The more information
that is provided, the better.

**Fixes issue #**:
Reference implementation for Version Management TAP (https://github.com/theupdateframework/taps/pull/107/files)

**Description of the changes being introduced by the pull request**:
This pull request adds a comparison of minor spec_versions (in format major.minor.fix) during an update and reports an out of date client to that client. Minor versions will not have breaking changes, so the update may continue.

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

### CODE REVIEW

**Review by lukpueh — COMMENTED** [review]
It might be worth taking a look at version parsing libraries? [`distutils.version`](https://docs.python.org/3/distutils/apiref.html#module-distutils.version) is in the standard library and should do the trick.

Or the third-party [`packaging.version`](https://packaging.pypa.io/en/latest/version/#version-handling), if [PEP-440](https://www.python.org/dev/peps/pep-0440/) compliance is required.

**Comment by trishankatdatadog** [comment]
@lukpueh https://github.com/DataDog/integrations-core/blob/23ebd01428da80b85ad59d5c7e23be57a1970cee/datadog_checks_downloader/datadog_checks/downloader/download.py#L298-L299

**Comment by awwad** [comment]
What's the goal here?  Existing code on the develop branch already does this comparison.  This adds some temp variables for minor and fix, only if they exist, and then checks them (whether or not they existed).

**Comment by mnm678** [comment]
I updated it to use parse_version which eliminates the temp variables and makes it a bit cleaner.

The goal is to report to the user if the minor versions do not match. The update can still continue if the major versions are the same, but this allows users to know when their client is out of date. This is a small change to make the reference implementation consistent with a proposed TAP (https://github.com/theupdateframework/taps/pull/107/files)

File: tuf/client/updater.py [hunk 1]
```diff
@@ -1495,14 +1496,21 @@ def _get_metadata_file(self, metadata_role, remote_filename,
           metadata_spec_version = metadata_signable['signed']['spec_version']
           metadata_spec_major_version = int(metadata_spec_version.split('.')[0])
           code_spec_major_version = int(tuf.SPECIFICATION_VERSION.split('.')[0])
-
+
```
Comment (lukpueh) [inline tuf/client/updater.py]: Please remove trailing whitespace

Comment (lukpueh) [reply tuf/client/updater.py]: OT: We might want to activate a corresponding lint policy (see [in-toto's pylintrc](https://github.com/in-toto/in-toto/blob/develop/pylintrc#L59)).

File: tuf/client/updater.py [hunk 2]
```diff
@@ -1495,14 +1496,21 @@ def _get_metadata_file(self, metadata_role, remote_filename,
...
                 repr(code_spec_major_version) + '; however, the obtained '
                 'metadata lists version number: ' + str(metadata_spec_version))
 
+          #report to user if minor versions do not match, continue with update
+          if pkg_resources.parse_version(metadata_spec_version) != pkg_resources.parse_version(tuf.SPECIFICATION_VERSION):
+            logger.info("Downloaded metadata that specifies a higher minor " +
+                "spec_version. This code has version " + str(tuf.SPECIFICATION_VERSION) +
+                "and the metadata lists version number " + str(metadata_spec_version) +
+                ". The update will continue as the major versions match.")
+
```
Comment (lukpueh) [inline tuf/client/updater.py:1519]: Please try to [wrap lines at character 79](https://github.com/secure-systems-lab/code-style-guidelines#maximum-line-length).

Comment (lukpueh) [reply tuf/client/updater.py:1519]: OT: This is actually something we should have the linter check as well.

File: tuf/client/updater.py [hunk 3]
```diff
@@ -1495,14 +1496,21 @@ def _get_metadata_file(self, metadata_role, remote_filename,
...
+          
           if metadata_spec_major_version != code_spec_major_version:
             raise tuf.exceptions.UnsupportedSpecificationError(
                 'Downloaded metadata that specifies an unsupported '
                 'spec_version.  This code supports major version number: ' +
                 repr(code_spec_major_version) + '; however, the obtained '
                 'metadata lists version number: ' + str(metadata_spec_version))
 
+          #report to user if minor versions do not match, continue with update
+          if pkg_resources.parse_version(metadata_spec_version) != pkg_resources.parse_version(tuf.SPECIFICATION_VERSION):
```
Comment (lukpueh) [inline tuf/client/updater.py]: If you compare the full versions you don't need to parse them. But if I understand correctly, you don't want to compare the full version, just the minor version.  The major version you already compare above and patch/fix version you don't care about.

Also, your log message says that one version (part) is higher than the other, so you will need to compare with `>` operator. 

I suggest you parse both `metadata_spec_version` and `tuf.SPECIFICATION_VERSION` only once and then use the major parts for the first (existing) check and the minor parts for the second (your new) check.

**Review by lukpueh — CHANGES_REQUESTED** [review]
Thanks for the patch. See my comments inline.

Please also consider reading our [dev guidelines](https://github.com/secure-systems-lab/lab-guidelines/blob/master/dev-workflow.md), especially the parts about uniform descriptive commit messages. They are very helpful for reviewers. :)

Thanks!

File: tuf/client/updater.py [hunk 4]
```diff
@@ -1493,16 +1493,30 @@ def _get_metadata_file(self, metadata_role, remote_filename,
...
+          code_spec_minor_version = int(code_spec_version_split[1])
+ 
           if metadata_spec_major_version != code_spec_major_version:
             raise tuf.exceptions.UnsupportedSpecificationError(
                 'Downloaded metadata that specifies an unsupported '
                 'spec_version.  This code supports major version number: ' +
                 repr(code_spec_major_version) + '; however, the obtained '
                 'metadata lists version number: ' + str(metadata_spec_version))
 
+          #report to user if minor versions do not match, continue with update
```
Comment (lukpueh) [inline tuf/client/updater.py:1511]: The comment suggests `!=` but the code and the log message `>`. Which one is it?

File: tuf/client/updater.py [hunk 5]
```diff
@@ -1493,16 +1493,30 @@ def _get_metadata_file(self, metadata_role, remote_filename,
...
+          if metadata_spec_minor_version > code_spec_minor_version:
+            logger.info("Downloaded metadata that specifies a higher minor " +
+                "spec_version. This code has version " + 
+                str(tuf.SPECIFICATION_VERSION) + 
+                "and the metadata lists version number " +
```
Comment (lukpueh) [inline tuf/client/updater.py]: I think you are missing a whitespace before `and`.

**Review by lukpueh — COMMENTED** [review]
Thanks or addressing my comments. I have two more minor concerns. Also you have trailing whitespace in  three lines, consider taking a look at https://github.com/secure-systems-lab/code-style-guidelines/issues/5, which describes how to detect them.

File: tuf/client/updater.py (hunk 4 above)
Comment (mnm678) [reply tuf/client/updater.py:1511]: I updated the operator and message to report any mismatch in the minor version.

**Review by lukpueh — APPROVED** [review]
LGTM! Please run the necessary git commands (e.g. as [suggested by probot](https://github.com/theupdateframework/tuf/pull/854/checks?check_run_id=144156013)) to "sign them off" retroactively.

**Pull request merged**